sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CbdeHandlerRule.cs

```diff
@@ -0,0 +1,106 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using SonarAnalyzer.Helpers;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using System.Collections.Immutable;
+using System.Linq;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Diagnostics.CodeAnalysis;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(S3949DiagnosticId)]
+    public sealed class CbdeHandlerRule : SonarDiagnosticAnalyzer
+    {
+        // Note: For now, this rule actually runs only under windows and outside of SonarLint
+        private readonly bool unitTest;
+
+        private const string S3949DiagnosticId = "S3949";
+        private const string S3949MessageFormat = "{0}";
+        private static readonly DiagnosticDescriptor ruleS3949 = DiagnosticDescriptorBuilder.GetDescriptor(S3949DiagnosticId, S3949MessageFormat, RspecStrings.ResourceManager, fadeOutCode: true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(ruleS3949);
+
+        private readonly ImmutableDictionary<string, DiagnosticDescriptor> ruleIdToDiagDescriptor = ImmutableDictionary<string, DiagnosticDescriptor>.Empty
+            .Add("S3949", ruleS3949);
+        public CbdeHandlerRule() : this(false) {}
+        private CbdeHandlerRule(bool unitTest)
+        {
+            this.unitTest = unitTest;
+        }
+        public static CbdeHandlerRule MakeUnitTestInstance()
+        {
+            return new CbdeHandlerRule(true);
+        }
+        protected sealed override void Initialize(SonarAnalysisContext context)
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                CbdeHandler cbdeHandler = new CbdeHandler(context, OnCbdeIssue, ShouldRunCbdeInContext);
+            }
+        }
+
+        // This functions is called for each issue found by cbde after it runs
+        private void OnCbdeIssue(String key, String message, Location loc, CompilationAnalysisContext context)
+        {
+            if (!ruleIdToDiagDescriptor.ContainsKey(key))
+                throw new InvalidOperationException($"CBDE should not raise issues on key {key}");
+
+            context.ReportDiagnosticWhenActive(Diagnostic.Create(ruleIdToDiagDescriptor[key], loc, message));
+        }
+
+        internal const string ConfigurationAdditionalFile = "ProjectOutFolderPath.txt";
+
+        [ExcludeFromCodeCoverage]
+        internal static bool IsProjectOutput(AdditionalText file) =>
+           ParameterLoader.ConfigurationFilePathMatchesExpected(file.Path, ConfigurationAdditionalFile);
+
+        // The logic used here is based on detecting a file which is present only when not run from SonarLint
+        // The logic is copied from FirstOrDefault(IsProjectOutput)
```

Comment:
could you add a `/// <see cref=/>` comment to point to the exact namespace+class+member you copied this from?